### PR TITLE
HIVE-27733: ensure PerfLogger is thread-safe;

### PR DESCRIPTION
**Pull request by henrib**
### What changes were proposed in this pull request?
Make the PerfLogger class thread-safe.

### Why are the changes needed?
Avoid an unpredictable ConcurrentModifcationException

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Unit

### CODE REVIEW

File: common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  @Test
+  public void testMT() throws InterruptedException {
+    final PerfLogger pl = PerfLogger.getPerfLogger(null, true);
+    // we run concurrently the getEndTimes and perfLogBegin/perfLogEnd:
+    // on a Mac M1, this test fails easily if the perflooger maps are hashmaps
+    ExecutorService executorService = Executors.newFixedThreadPool(64);
+    // An executing threads counter
+    AtomicInteger count = new AtomicInteger(0);
+    // getEndTimes in a loop
+    executorService.execute(new Runnable() {
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java:57]: can use lambda
```
    executorService.execute(() -> {
      PerfLogger.setPerfLogger(pl);
      try {
        count.incrementAndGet();
        snooze(100);
        for (int i = 0; i < 64; ++i) {
          snooze(50);
          Map<String, Long> et = pl.getEndTimes();
          Assert.assertNotNull(et);
        }
      } finally {
        count.decrementAndGet();
        synchronized (count) {
          count.notifyAll();
        }
      }
```

File: common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          count.decrementAndGet();
+          synchronized (count) {
+            count.notifyAll();
+          }
+        }
+      }
+    });
+    // 32 threads calling perLogBeing/perfLogEnd
+    for(int t = 0; t < 31; ++t) {
+      executorService.execute(new Runnable() {
```
Comment (ayushtkn) [inline common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java:78]: can use lambda

**Review by ayushtkn — COMMENTED** [review]
Thanx @henrib, makes sense. dropped some comments.
Curious as in which case does this occur in actual query?

File: common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java [hunk 3]
```diff
@@ -89,17 +90,13 @@ public class PerfLogger {
...
-  static final private Logger LOG = LoggerFactory.getLogger(PerfLogger.class.getName());
-  protected static final ThreadLocal<PerfLogger> perfLogger = new ThreadLocal<PerfLogger>();
+  private static final Logger LOG = LoggerFactory.getLogger(PerfLogger.class.getName());
+  protected static final ThreadLocal<PerfLogger> perfLogger = new ThreadLocal<>();
 
 
-  private PerfLogger() {
-    // Use getPerfLogger to get an instance of PerfLogger
-  }
-
```
[the remark this answers is not in the document]
Comment (henrib) [reply common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java:102]: No good reason, putting it back.

File: common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java [hunk 4]
```diff
@@ -131,7 +128,9 @@ public static void setPerfLogger(PerfLogger resetPerfLogger) {
   public void perfLogBegin(String callerName, String method) {
     long startTime = System.currentTimeMillis();
     startTimes.put(method, Long.valueOf(startTime));
-    LOG.debug("<PERFLOG method={} from={}>", method, callerName);
+    if (LOG.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (henrib) [reply common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java:131]: It is not necessary (force of habit I guess) plus it makes the code similar to what's going on in perfLogEnd.

File: common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java [hunk 5]
```diff
@@ -177,21 +176,13 @@ public long perfLogEnd(String callerName, String method, String additionalInfo)
...
 
   public Long getEndTime(String method) {
-    long endTime = 0L;
-
-    if (endTimes.containsKey(method)) {
-      endTime = endTimes.get(method);
-    }
-    return endTime;
+    Long time = endTimes.get(method);
+    return time != null? time.longValue() : 0L;
```
[the remark this answers is not in the document]
Comment (henrib) [reply common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java:185]: Will do.

File: common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java [hunk 6]
```diff
@@ -177,21 +176,13 @@ public long perfLogEnd(String callerName, String method, String additionalInfo)
...
 
   public Long getStartTime(String method) {
-    long startTime = 0L;
-
-    if (startTimes.containsKey(method)) {
-      startTime = startTimes.get(method);
-    }
-    return startTime;
+    Long time = startTimes.get(method);
+    return time != null? time.longValue() : 0L;
```
[the remark this answers is not in the document]
Comment (henrib) [reply common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java:180]: We can and I will update some signatures accordingly.

File: common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java [hunk 7]
```diff
@@ -203,11 +194,12 @@ public boolean endTimeHasMethod(String method) {
   }
 
   public Long getDuration(String method) {
-    long duration = 0;
-    if (startTimes.containsKey(method) && endTimes.containsKey(method)) {
-      duration = endTimes.get(method) - startTimes.get(method);
+    Long startTime = startTimes.get(method);
+    Long endTime = endTimes.get(method);
+    if (startTime != null && endTime != null) {
+      return endTime.longValue() - startTime.longValue();
```
[the remark this answers is not in the document]
Comment (henrib) [reply common/src/java/org/apache/hadoop/hive/ql/log/PerfLogger.java:200]: Yes and signature does not really need to be a Long (vs long).

File: common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java (hunk 1 above)
Comment (henrib) [reply common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java:57]: Done.

File: common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java (hunk 2 above)
Comment (henrib) [reply common/src/test/org/apache/hadoop/hive/ql/log/PerfLoggerTest.java:78]: Done.

**Comment by henrib** [comment]
> Thanx @henrib, makes sense. dropped some comments.
> Curious as in which case does this occur in actual query?

It is an elusive problem that only occurs sporadically in some tests; the concurrency window triggering the error seems very small so it is not easy to reproduce. But the problem does occur...

**Review by ayushtkn — APPROVED** [review]
LGTM

**Comment by ramesh0201** [comment]
@ayushtkn Can this PR be merged?

**Pull request merged**